model/src/test/kotlin/AbstractDependencyNavigatorTest.kt

```diff
@@ -0,0 +1,360 @@
+/*
+ * Copyright (C) 2021 Bosch.IO GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.model
+
+import io.kotest.core.spec.style.WordSpec
+import io.kotest.matchers.collections.beEmpty
+import io.kotest.matchers.collections.contain
+import io.kotest.matchers.collections.containAll
+import io.kotest.matchers.collections.containExactly
+import io.kotest.matchers.collections.containExactlyInAnyOrder
+import io.kotest.matchers.collections.haveSize
+import io.kotest.matchers.collections.shouldContainExactlyInAnyOrder
+import io.kotest.matchers.should
+import io.kotest.matchers.shouldBe
+import io.kotest.matchers.shouldNot
+
+import java.io.File
+import java.time.Instant
+
+import org.ossreviewtoolkit.utils.test.readOrtResult
+
+/**
+ * A base class for tests of concrete [DependencyNavigator] implementations.
+ *
+ * The class is configured with an ORT result file that contains the expected results in a specific format. It then
+ * runs tests on the [DependencyNavigator] of this result and checks whether it returns the correct dependency
+ * information.
+ */
+abstract class AbstractDependencyNavigatorTest : WordSpec() {
+    /** The name of the result file to be used by all test cases. */
+    protected abstract val resultFileName: String
+
+    /**
+     * The name of the file with a result that contains issues. This is used by tests of the collectIssues() function.
+     */
+    protected abstract val resultWithIssuesFileName: String
+
+    private val testResult by lazy { readOrtResult(resultFileName) }
+
+    private val testProject by lazy { testResult.getProject(PROJECT_ID)!! }
+
+    protected val navigator by lazy { testResult.dependencyNavigator }
+
+    init {
+        "scopeNames" should {
+            "return the scope names of a project" {
+                navigator.scopeNames(testProject) should containExactlyInAnyOrder("compile", "test")
+            }
+        }
+
+        "directDependencies" should {
+            "return the direct dependencies of a project" {
+                navigator.directDependencies(testProject, "test")
+                    .map { it.id }.toList() should containExactly(
```

Comment:
Nit: There should be no break between these lines to fix indentation.

Reply:
You can also avoid the `.toList()` conversion by importing the specialized `io.kotest.matchers.sequences.containExactly` (eventually under a different name, if you also need `containExactly` for collections).

Reply:
I tried this, but then, surprisingly, the test failed. After a longer investigation, I think the problem is in the way the Kotest matcher compares sequences. The sequence with dependencies is constructed using `generateSequence()` and backed by a function that calculates the next element. It can therefore be traversed only once (which I would assume is true for most of the sequences generated this way).

The sequence matcher, however, first determines the count of the sequence (which requires a traversal) and afterwards checks the elements (another traversal). This second traversal no longer returns the full sequence, causing the comparison to fail.

Reply:
@sksamuel acknowledged it's an [issue](https://github.com/kotest/kotest/issues/2412).